From f43ae8e10378b5e078b168f26e729c1044802865 Mon Sep 17 00:00:00 2001
From: Abhinav Kumar <quic_abhinavk@quicinc.com>
Date: Sat, 29 Nov 2025 16:19:42 +0800
Subject: [PATCH 3/6] FROMLIST: drm/msm/dp: add support for programming
 p1/p2/p3 register block

QCS8300 support 4 streams MST and SA8775P support 2 x 2 MST. Add
support for other 3 pixels register block, and introduced
msm_dp_read_pn/msm_dp_write_pn to minimize repetitive code. All PCLK
share the same register definitions with different base addresses.

Signed-off-by: Abhinav Kumar <quic_abhinavk@quicinc.com>
Signed-off-by: Yongxing Mou <quic_yongmou@quicinc.com>
Upstream-Status: Submitted [https://lore.kernel.org/all/20250609-msm-dp-mst-v2-12-a54d8902a23d@quicinc.com/]
---
 drivers/gpu/drm/msm/dp/dp_catalog.c | 50 +++++++++++++++++++++++++++++
 drivers/gpu/drm/msm/dp/dp_ctrl.c    | 12 +++++++
 2 files changed, 62 insertions(+)

diff --git a/drivers/gpu/drm/msm/dp/dp_catalog.c b/drivers/gpu/drm/msm/dp/dp_catalog.c
index dfd46e1f630e..bf6f477a27e5 100644
--- a/drivers/gpu/drm/msm/dp/dp_catalog.c
+++ b/drivers/gpu/drm/msm/dp/dp_catalog.c
@@ -79,6 +79,8 @@ struct dss_io_data {
 	struct dss_io_region link;
 	struct dss_io_region p0;
 	struct dss_io_region p1;
+	struct dss_io_region p2;
+	struct dss_io_region p3;
 };
 
 struct msm_dp_catalog_private {
@@ -173,6 +175,46 @@ static inline u32 msm_dp_read_p1(struct msm_dp_catalog_private *catalog,
 	return readl_relaxed(catalog->io.p1.base + offset);
 }
 
+static inline void msm_dp_write_p2(struct msm_dp_catalog_private *catalog,
+				   u32 offset, u32 data)
+{
+	/*
+	 * To make sure interface reg writes happens before any other operation,
+	 * this function uses writel() instread of writel_relaxed()
+	 */
+	writel(data, catalog->io.p2.base + offset);
+}
+
+static inline u32 msm_dp_read_p2(struct msm_dp_catalog_private *catalog,
+				 u32 offset)
+{
+	/*
+	 * To make sure interface reg writes happens before any other operation,
+	 * this function uses writel() instread of writel_relaxed()
+	 */
+	return readl_relaxed(catalog->io.p2.base + offset);
+}
+
+static inline void msm_dp_write_p3(struct msm_dp_catalog_private *catalog,
+				   u32 offset, u32 data)
+{
+	/*
+	 * To make sure interface reg writes happens before any other operation,
+	 * this function uses writel() instread of writel_relaxed()
+	 */
+	writel(data, catalog->io.p3.base + offset);
+}
+
+static inline u32 msm_dp_read_p3(struct msm_dp_catalog_private *catalog,
+				 u32 offset)
+{
+	/*
+	 * To make sure interface reg writes happens before any other operation,
+	 * this function uses writel() instread of writel_relaxed()
+	 */
+	return readl_relaxed(catalog->io.p3.base + offset);
+}
+
 static inline u32 msm_dp_read_link(struct msm_dp_catalog_private *catalog, u32 offset)
 {
 	return readl_relaxed(catalog->io.link.base + offset);
@@ -1387,6 +1429,14 @@ static int msm_dp_catalog_get_io(struct msm_dp_catalog_private *catalog)
 		dss->p1.base = msm_dp_ioremap(pdev, 4, &dss->p1.len);
 		if (IS_ERR(dss->p1.base))
 			DRM_DEBUG("unable to remap p1 region: %pe\n", dss->p1.base);
+
+		dss->p2.base = msm_dp_ioremap(pdev, 5, &dss->p2.len);
+		if (IS_ERR(dss->p2.base))
+			DRM_DEBUG("unable to remap p2 region: %pe\n", dss->p2.base);
+
+		dss->p3.base = msm_dp_ioremap(pdev, 6, &dss->p3.len);
+		if (IS_ERR(dss->p3.base))
+			DRM_DEBUG("unable to remap p3 region: %pe\n", dss->p3.base);
 	}
 
 	return 0;
diff --git a/drivers/gpu/drm/msm/dp/dp_ctrl.c b/drivers/gpu/drm/msm/dp/dp_ctrl.c
index df43017c689e..930e5a7e96f4 100644
--- a/drivers/gpu/drm/msm/dp/dp_ctrl.c
+++ b/drivers/gpu/drm/msm/dp/dp_ctrl.c
@@ -2439,6 +2439,18 @@ static int msm_dp_ctrl_clk_init(struct msm_dp_ctrl *msm_dp_ctrl)
 	else
 		ctrl->num_pixel_clks++;
 
+	ctrl->pixel_clk[DP_STREAM_2] = devm_clk_get(dev, "stream_2_pixel");
+	if (IS_ERR(ctrl->pixel_clk[DP_STREAM_2]))
+		DRM_DEBUG("failed to get stream_2_pixel clock");
+	else
+		ctrl->num_pixel_clks++;
+
+	ctrl->pixel_clk[DP_STREAM_3] = devm_clk_get(dev, "stream_3_pixel");
+	if (IS_ERR(ctrl->pixel_clk[DP_STREAM_3]))
+		DRM_DEBUG("failed to get stream_3_clock clock");
+	else
+		ctrl->num_pixel_clks++;
+
 	return 0;
 }
 
-- 
2.34.1

